lib/sass/script/functions.rb

```diff
@@ -2078,17 +2089,61 @@ def map_get(map, key)
     # @example
     #   map-merge(("foo": 1), ("bar": 2)) => ("foo": 1, "bar": 2)
     #   map-merge(("foo": 1, "bar": 2), ("bar": 3)) => ("foo": 1, "bar": 3)
-    # @overload map_merge($map1, $map2)
+    #   map-merge((a: (b: c)), a, (d: e)) => (a: (b: c, d: e))
+    #   map-merge((a: (b: c)), (a: (d: e))) => (a: (d: e))
+    # @overload map_merge($map1, $keys..., $map2)
     #   @param $map1 [Sass::Script::Value::Map]
+    #   @param $keys [[Sass::Script::Value]]
     #   @param $map2 [Sass::Script::Value::Map]
     # @return [Sass::Script::Value::Map]
     # @raise [ArgumentError] if either parameter is not a map
-    def map_merge(map1, map2)
+    def map_merge(map1, *keys)
       assert_type map1, :Map, :map1
+
+      # Empty Maps arrive as Lists, so ensure we are a list if we are empty
+      # but since Lists don't define the merge functions, we need to ensure
+      # we have a proper Map.
+      map1 = map(map1.to_h)
+
+      # The last argument is always the map we are merging in
+      map2 = keys.pop
+
+      # Make sure that the last argument is Map-like, however it
+      # can be a List, and we're fine with that.
       assert_type map2, :Map, :map2
-      map(map1.to_h.merge(map2.to_h))
+
+      map1.recursive_merge(keys, map2)
     end
     declare :map_merge, [:map1, :map2]
+    declare :map_merge, [:map1], :var_args => true
+
+    # Allows the setting of specific values in a
+    # map by passing in a series of keys, with the
+    # final argument setting the value of that key.
+    #
+    # If no keys exist, they are created along with
+    # any needed nested maps.
+    #
+    # @example
+    #   map-set((a: 1), b, 2) => (a: 1, b: 2)
+    #   map-set((a: (b: 1)), a, c, 2) => (a: (b: 1, c: 2))
+    #   map-set((), a, b, c) => (a: (b: c))
+    # @overload map-set($map, $keys..., $value)
+    #   @param $map1 [Sass::Script::Value::Map]
+    #   @param $keys [[Sass::Script::Value::Base]]
+    #   @param $value [Sass::Script::Value::Base]
+    # @return [Sass::Script::Value::Map]
+    # @raise [ArgumentError] if the first parameter is not a map
+    def map_set(map, *keys)
+      assert_type map, :Map, :map
+      new_value = keys.pop
+      if map.is_a?(Sass::Script::Value::List)
+        map = map(map.to_h)
+      end
+      map.recursive_set(keys, new_value)
+    end
+    declare :map_set, [:map, :value]
+    declare :map_set, [:map, :keys], :var_args => true
```

Comment:
The names of var args should never appear in declarations. It's not valid to pass var args by name in Sass.